### ISSUE TITLE
A way to expand all groups by default?

### ISSUE THREAD

**Issue by guidedways** [h6ah4i/android-advancedrecyclerview#281]
Right now I end up expanding each group manually. Is there a way to do this at layout time? Perhaps a flag that you can set on the Layout Manager?

**Comment by h6ah4i** [comment]
@guidedways Hi. Use the `RecyclerViewExpandableItemManager.expandAll()` method. [Please check the demo app for more details](https://github.com/h6ah4i/android-advancedrecyclerview/blob/0.9.2/example/src/main/java/com/h6ah4i/android/example/advrecyclerview/demo_e_already_expanded/AlreadyExpandedGroupsExpandableExampleFragment.java#L81). Thanks.

**Comment by guidedways** [comment]
That's not so great as there can be a delay between the time notifyDataChanged() and the time it actually finishes laying out. I usually end up seeing glitches, which is why it would be ideal to have all groups expanded by default without having to do this as a second step.

**Comment by h6ah4i** [comment]
Hi. Do you call `expandAll()` method before setting `setAdapter()`? In this case, `notifyDataChanged()` method performs nothing because no observers are attached yet.

If you want to add group item(s) expanded, use `RecyclerViewExpandableItemManager.notifyGroupItemInserted(int groupPosition, boolean expanded)` or `RecyclerViewExpandableItemManager.notifyGroupItemRangeInserted(int groupPositionStart, int itemCount, boolean expanded)`. The last parameter of these methods can control expanded state of newly added items.


If all of above not fit to your situation, I guess adding a new method like the following will solve the problem. Is this what you need? This method performs `RecyclerView.Adapter.notifyDataSetChanged()` and `expandGroup()`/`collapseGroup()` method atomically.

```java
void RecyclerViewExpandableItemManager.notifyDataSetChanged(int expandedStateControl);

when expandedStateControl = 0: preserve current expanded state
when expandedStateControl = 1: expand all groups
when expandedStateControl = 2: collapse all groups
```

**Comment by guidedways** [comment]
Thanks - sadly the first option won't work as I need to call `notifyDataChanged()` directly most of the times, as too much changes in the background based on various factors.

I think what I really want is the opposite of the method you're proposing; I need all groups to default to 'expanded' unless I collapse them by code or otherwise once the data has finished loading. What I've seen for now in many situations is that I try and 'expand' a group after calling `notifyDataChanged()` and it does a half-assed animation and then ends up still collapsed. Even the `checkExpandedState` callbacks are not called. Ideally the best solution would be to have the `checkExpandedState` (can't recall the exact method name) method called on the Adapter data is changed. At least at that point we could avoid weird animations which should not execute after data changes. It should just expand / collapse based on what the adapter says.

So in short, I'd prefer the adapter callback gets called at the point notifyDataChanged is called, or at least have a method on the layout manager so I can default everything to remain expanded unless I collapse it at some later point. After all, it doesn't make sense to have groups in their collapsed state in most cases, unless you're developing a Folder / File viewer.

**Comment by h6ah4i** [comment]
@guidedways Hi. Thank you for suggestion. I've understood the importance of having the default state whether groups are expanded or not. I added the `setDefaultGroupsExpandedState()` method to `RecyclerViewExpandableItemManager`. Could you test the new method work in your project?


**Usage:**
```java
expMgr = new RecyclerViewExpandableItemManager(eimSavedState);

// Expand all group items by default. This method must be called before creating a wrapper adapter.
expMgr.setDefaultGroupsExpandedState(true);

expMgr.createWrappedAdapter(myItemAdapter); 
```

**Gradle setting:**
```gradle
repositories {
    maven { url "https://jitpack.io" }
}

dependencies {
    compile 'com.github.h6ah4i:android-advancedrecyclerview:7d4ec2a86e9eb908a249e212f6b3facc23053672'
}
```


---

>  Even the checkExpandedState callbacks are not called. Ideally the best solution would be to have the checkExpandedState (can't recall the exact method name) method called on the Adapter data is changed. 

Hmm... I believe this library does not provide such callbacks.


> So in short, I'd prefer the adapter callback gets called at the point notifyDataChanged is called, or at least have a method on the layout manager so I can default everything to remain expanded unless I collapse it at some later point. After all, it doesn't make sense to have groups in their collapsed state in most cases, unless you're developing a Folder / File viewer.

Sorry, I do not understand what you mean. This library does not provide any layout managers for now :(  Do you mean "layout manager" = `RecyclerViewExpandableItemManager`?

I think layout managers should not take care about groups/children because `RecyclerView` can only manage a **flat** list. In this library, wrapper adapter is the what managing groups/children and its expanded state. The wrapper adapter acts as a normal `RecyclerView.Adapter` but it translate **expandable list to flat list** internally.

**Comment by JureVI** [comment]
Seems to be working here, thanks! I'll get back to you if I notice any strange behavior.

**Comment by ch-muhammad-adil** [comment]
Hello .. Any way to stop the expanding before. I want the views expanded but on click of parent view it should not collapse. Waiting for quick reply, I am in a trouble. Thanks,

**Comment by ch-muhammad-adil** [comment]
Thanks I found solution 

    @Override
    public boolean onCheckCanExpandOrCollapseGroup(MyGroupViewHolder holder, int groupPosition, int x, int y, boolean expand) {
        return false;
    }

returning false worked. :)

**Comment by guidedways** [comment]
Thanks - this is working fine. Any plans of pushing this into the stable release?

**Comment by h6ah4i** [comment]
Version 0.9.3 is out now. It includes the fix of this issue.
